### PR TITLE
[modbus] Modbus register array backed by bytes and other simplifications

### PR DESCRIPTION
**Pull request by ssalonen**
(this is equivalent to #8862 but against `main` branch)

Inspired by #8172 I reworked the modbus transport to use direct byte arrays to represent the data. 

This is purely refactoring with the aim of reducing boilerplate byte manipulation/copying code in bindings. With the changes we do get boosted performance: with 45000 invocations to `extractStateFromRegisters` with varying inputs, the self time is dropped to 55% of original, total time to 80%. Naturally these type of performance gains are dwarfed by network I/O and other real time overheads.

This has many benefits:
- Unnecessary boxing / unboxing of ModbusRegister objects is removed all-together and one can operate directly with familiar `byte[]` using bitwise operations
- Some modbus bindings constructed `byte[]` from `ModbusRegister[]` (essentially array of 16-bit data blocks, registers). 

In addition I brought in more general methods to work with the data:
- Methods such as `extractXX`, e.g. `extractSInt8` (extract 8-bit signed integer) working directly with `byte[]` data. Return type is directly smallest suitable java primitive such as `byte`, `short` etc.
- `ValueBuffer` interface that acts similar to java's `ByteBuffer` but having specialized methods to extract different types of value types that are encountered with modbus

Finally, simplifications & harmonization was made to how registers are converted to to numbers, and how commands are converted to bytes. I find the new conversion logic more readable.

The PR is splitted in two parts:
1. changes in transport API. Mainly backwards compatible. Only exception is extractStringFromXX which now returns `String`, not `StringType`
2. Updates and simplifications to individual bindings. Tests are added, if missing, to have regression tests for data type conversions

CC code owners to review at least the changes in the relevant binding

/bundles/org.openhab.binding.modbus.e3dc/ @weymann
/bundles/org.openhab.binding.modbus.helioseasycontrols/ @bern77
/bundles/org.openhab.binding.modbus.stiebeleltron/ @pail23
/bundles/org.openhab.binding.modbus.studer/ @giovannimirulla
/bundles/org.openhab.binding.modbus.sunspec/ @mrbig

Would be great if you could test with real devices as well

### CODE REVIEW

File: bundles/org.openhab.binding.modbus/pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
-    <version>3.0.0-SNAPSHOT</version>
+    <version>2.5.10-SNAPSHOT</version>
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/pom.xml:10]: fix

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/discovery/internal/ModbusDiscoveryService.java [hunk 2]
```diff
@@ -44,7 +44,7 @@
  * @author Nagy Attila Gabor - initial contribution
  *
  */
-@Component(service = DiscoveryService.class, configurationPid = "discovery.modbus")
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "discovery.modbus")
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/discovery/internal/ModbusDiscoveryService.java:47]: fix

File: bundles/org.openhab.io.transport.modbus/pom.xml [hunk 3]
```diff
@@ -7,7 +7,7 @@
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
-    <version>3.0.0-SNAPSHOT</version>
+    <version>2.5.10-SNAPSHOT</version>
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/pom.xml:10]: fix

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesCommandToRegistersTest.java [hunk 4]
```diff
@@ -307,24 +325,22 @@
     }
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
-    @ParameterizedTest
-    @MethodSource("data")
-    public void testCommandToRegisters(Command command, ValueType type, Object expectedResult) {
+    @Test
+    public void testCommandToRegisters() {
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesCommandToRegistersTest.java:329]: revert changed annotations

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesCommandToRegistersTest.java [hunk 5]
```diff
@@ -47,6 +64,7 @@
         return shorts;
     }
 
+    @Parameters
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesCommandToRegistersTest.java:67]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesExtractIndividualMethodsTest.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+    public BitUtilitiesExtractIndividualMethodsTest(Object expectedResult, ValueType type, byte[] bytes,
+            int byteIndex) {
+        this.expectedResult = expectedResult; // Exception or DecimalType
+        this.type = type;
+        this.bytes = bytes;
+        this.byteIndex = byteIndex;
+    }
+
+    @Parameters
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/BitUtilitiesExtractIndividualMethodsTest.java:58]: follow example set in other tests, removing this

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java [hunk 7]
```diff
@@ -118,9 +113,14 @@
 
     private static AtomicCounter udpServerIndex = new AtomicCounter(0);
 
-    protected @Spy TCPSlaveConnectionFactory tcpConnectionFactory = new TCPSlaveConnectionFactoryImpl();
-    protected @Spy UDPSlaveTerminalFactory udpTerminalFactory = new UDPSlaveTerminalFactoryImpl();
-    protected @Spy SerialConnectionFactory serialConnectionFactory = new SerialConnectionFactoryImpl();
+    @Spy
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java:116]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java [hunk 8]
```diff
@@ -65,8 +62,6 @@
 /**
  * @author Sami Salonen - Initial contribution
  */
-@ExtendWith(MockitoExtension.class)
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java:68]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java [hunk 9]
```diff
@@ -146,7 +146,7 @@ protected static InetAddress localAddress() throws UnknownHostException {
         return InetAddress.getByName("127.0.0.1");
     }
 
-    @BeforeEach
+    @Before
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/IntegrationTestSupport.java:149]: revert

File: itests/org.openhab.binding.modbus.tests/src/main/java/org/openhab/binding/modbus/tests/ModbusDataHandlerTest.java [hunk 10]
```diff
@@ -88,6 +88,7 @@
 /**
  * @author Sami Salonen - Initial contribution
  */
+@RunWith(MockitoJUnitRunner.class)
```
Comment (ssalonen) [inline itests/org.openhab.binding.modbus.tests/src/main/java/org/openhab/binding/modbus/tests/ModbusDataHandlerTest.java:91]: revert

File: itests/org.openhab.binding.modbus.tests/src/main/java/org/openhab/binding/modbus/tests/ModbusDataHandlerTest.java [hunk 11]
```diff
@@ -113,7 +114,7 @@ public String transform(String function, String source) throws TransformationExc
     }
     private List<ModbusWriteRequestBlueprint> writeRequests = new ArrayList<>();
 
-    @AfterEach
```
Comment (ssalonen) [inline itests/org.openhab.binding.modbus.tests/src/main/java/org/openhab/binding/modbus/tests/ModbusDataHandlerTest.java:116]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java [hunk 12]
```diff
@@ -84,22 +83,22 @@ public void testFC6SingleRegisterMaxTries99() {
...
+        WriteRequestJsonUtilities.fromJson(55, "[{"//
                 + "\"functionCode\": 6,"//
                 + "\"address\": 5412,"//
                 + "\"value\": [3, 4]"//
-                + "}]"));
+                + "}]");
     }
 
-    @Test
+    @Test(expected = IllegalArgumentException.class)
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java:95]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java [hunk 13]
```diff
@@ -84,22 +83,22 @@ public void testFC6SingleRegisterMaxTries99() {
                         (Matcher) new RegisterMatcher(55, 5412, 99, ModbusWriteFunctionCode.WRITE_SINGLE_REGISTER, 3)));
     }
 
-    @Test
+    @Test(expected = IllegalArgumentException.class)
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java:86]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java [hunk 14]
```diff
@@ -50,13 +49,13 @@ public void testEmptyArray() {
         assertThat(WriteRequestJsonUtilities.fromJson(3, "[]").size(), is(equalTo(0)));
     }
 
-    @Test
+    @Test(expected = IllegalArgumentException.class)
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java:52]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java [hunk 15]
```diff
@@ -25,7 +24,7 @@
 
 import org.eclipse.jdt.annotation.NonNull;
 import org.hamcrest.Matcher;
-import org.junit.jupiter.api.Test;
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java:28]: revert

File: bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java [hunk 16]
```diff
@@ -13,9 +13,8 @@
 package org.openhab.io.transport.modbus.test;
 
 import static org.hamcrest.CoreMatchers.*;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.collection.ArrayMatching.arrayContaining;
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/test/java/org/openhab/io/transport/modbus/test/WriteRequestJsonUtilitiesTest.java:17]: revert

**Comment by ssalonen** [comment]
JUnit changes need reverting

File: bundles/org.openhab.binding.modbus.helioseasycontrols/README.md [hunk 17]
```diff
@@ -283,58 +283,46 @@ The binding provides the following properties:
...
 
 // Standby mode
-Switch KWL_Standby_Mode                  "Standby mode"                              <fan_off>     (gKWL) {channel="modbus:helios-easycontrols:modbus-gateway:kwl:operation#standbyModeStatus"}
-Number KWL_Standby_Mode_Duration         "Standby mode duration"                     <clock>       (gKWL) {channel="modbus:helios-easycontrols:modbus-gateway:kwl:operation#standbyModeDuration"}
-Number KWL_Standby_Mode_Stage            "Standby mode fan stage"                    <fan>         (gKWL) {channel="modbus:helios-easycontrols:modbus-gateway:kwl:operation#standbyModeFanStage"}
-Number KWL_Standby_Mode_Remaining        "Standby mode remaining time [%d min]"      <clock>       (gKWL) {channel="modbus:helios-easycontrols:modbus-gateway:kwl:operation#standbyModeRemainingTime"}
+Switch KWL_Standby_Mode                  "Standby mode"                              <fan_off>     (gKWL) {channel="modbus:helios-ventilation-easycontrols:modbus-gateway:kwl:operation#standbyModeStatus"}
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/README.md:305]: Revert

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java [hunk 18]
```diff
@@ -47,60 +47,80 @@ public void setThingHandler(@Nullable ThingHandler handler) {
 
     private void triggerSwitch(String variableName) {
         try {
-            if (handler != null) {
-                handler.writeValue(variableName, "1");
+            if (this.handler != null) {
+                this.handler.writeValue(variableName, "1");
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:51]: Revert

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java [hunk 19]
```diff
@@ -47,60 +47,80 @@ public void setThingHandler(@Nullable ThingHandler handler) {
...
-    @RuleAction(label = "reset filter change timer", description = "Sets the filter change timer back to the configured interval.")
+    @RuleAction(label = "Reset filter change timer", description = "Sets the filter change timer back to the configured interval")
     public void resetFilterChangeTimer() {
-        triggerSwitch(HeliosEasyControlsBindingConstants.FILTER_CHANGE_RESET);
+        this.triggerSwitch(HeliosEasyControlsBindingConstants.FILTER_CHANGE_RESET);
     }
 
-    public static void resetFilterChangeTimer(ThingActions actions) {
-        ((HeliosEasyControlsActions) actions).resetFilterChangeTimer();
+    public static void resetFilterChangeTimer(@Nullable ThingActions actions) {
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:63]: Revert

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java [hunk 20]
```diff
@@ -47,60 +47,80 @@ public void setThingHandler(@Nullable ThingHandler handler) {
...
         } catch (HeliosException e) {
             logger.warn("Error executing action 'resetFilterChangeTimer': {}", e.getMessage());
         }
     }
 
-    @RuleAction(label = "reset filter change timer", description = "Sets the filter change timer back to the configured interval.")
+    @RuleAction(label = "Reset filter change timer", description = "Sets the filter change timer back to the configured interval")
     public void resetFilterChangeTimer() {
-        triggerSwitch(HeliosEasyControlsBindingConstants.FILTER_CHANGE_RESET);
+        this.triggerSwitch(HeliosEasyControlsBindingConstants.FILTER_CHANGE_RESET);
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:60]: Revert

File: bundles/org.openhab.binding.modbus.studer/src/main/feature/feature.xml [hunk 21]
```diff
@@ -3,7 +3,7 @@
 	<repository>file:${basedirRoot}/bundles/org.openhab.io.transport.modbus/target/feature/feature.xml</repository>
 	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
 
-	<feature name="openhab-binding-modbus-studer" description="Studer Binding" version="${project.version}">
+	<feature name="openhab-bindings-modbus-studer" description="Studer Binding" version="${project.version}">
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.studer/src/main/feature/feature.xml:6]: Revert

File: bundles/org.openhab.binding.modbus.sunspec/NOTICE [hunk 22]
```diff
@@ -10,4 +10,4 @@ https://www.eclipse.org/legal/epl-2.0/.
 
 == Source Code
 
-https://github.com/openhab/openhab-addons
+https://github.com/openhab/openhab2-addons
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sunspec/NOTICE:13]: Revert

File: bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryParticipant.java [hunk 23]
```diff
@@ -33,7 +33,7 @@
  * @author Nagy Attila Gabor - initial contribution
  *
  */
-@Component
+@Component(immediate = true)
```
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryParticipant.java:36]: Revert

**Comment by weymann** [comment]
Sorry, I don't get it! I see you want to merge your openhab**2**-addons merged into main but main is intended only for 3.0 or am I wrong?

**Comment by ssalonen** [comment]
@weymann we live in interesting times with openHAB3 coming up :) 

This PR is in fact [auto-migrated](https://github.com/openhab/openhab-addons/pull/8862#issuecomment-716137007) version of OH2 branch #8862. [Per guidance from @fwolter](https://github.com/openhab/openhab-addons/pull/8862#issuecomment-716133365), `main` should primary target for new any new PRs...I have understood only few selected parameters are back-ported to OH2.

Don't get confused by my repo name....

Since the code has been auto-migrated, I need to do some additional manual corrections (similar to [PR #8519](https://github.com/openhab/openhab-addons/pull/8519)) still related to unit tests (JUnit version is newer with OH3, and it is used in different way). Similarly, some `feature.xml` and `NOTICE` files should remain unchanged in this PR (commented above as well).

Probably it makes sense that I do above corrections before real tests are proceeded with.

**Comment by mrbig** [comment]
I've checked the changes in sunspec, and they are all right. I like the simplified API, thans @ssalonen

**Comment by weymann** [comment]
Ok, so I'll test it on 2.5. 

I'm able to build the changes. But I don't know how to replace io.transport.modbus so I get the error below. Can you give some advice how to exchange this bundle?

```
2020-10-27 16:55:51.490 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.NoSuchMethodError: org.openhab.io.transport.modbus.ModbusRegisterArray.getBytes()[B
	at org.openhab.binding.modbus.e3dc.internal.modbus.Parser.handle(Parser.java:64) ~[?:?]
```

**Comment by ssalonen** [comment]
Good question @weymann ...

Adapting instructions here might help out https://community.openhab.org/t/modbus-stops-without-any-errors/104050/38?u=ssalonen

File: bundles/org.openhab.binding.modbus.helioseasycontrols/README.md (hunk 17 above)
Comment (bern77) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/README.md:305]: Actually the changes you would revert here were intentional as per my last PR (#8804), i.e. the 'ventilation' part in the thing type is gone.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java (hunk 18 above)
Comment (bern77) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:51]: These changes were the result of  #8755 and should therefore remain.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java (hunk 19 above)
Comment (bern77) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:63]: These changes were the result of  #8755 and should therefore remain.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java (hunk 20 above)
Comment (bern77) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:60]: These changes were the result of  #8755 and should therefore remain.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/pom.xml [hunk 24]
```diff
@@ -7,7 +7,7 @@
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
-    <version>3.0.0-SNAPSHOT</version>
+    <version>2.5.10-SNAPSHOT</version>
```
Comment (bern77) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/pom.xml:10]: I thought this PR should be for OH3?

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/feature/feature.xml [hunk 25]
```diff
@@ -2,7 +2,7 @@
 <features name="org.openhab.binding.modbus.helioseasycontrols-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
 	<repository>file:${basedirRoot}/bundles/org.openhab.io.transport.modbus/target/feature/feature.xml</repository>
 
-	<feature name="openhab-binding-modbus-helioseasycontrols" description="Modbus.HeliosEasyControls Binding" version="${project.version}">
+	<feature name="openhab-binding-modbus.helioseasycontrols" description="Modbus.HeliosEasyControls Binding" version="${project.version}">
```
Comment (bern77) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/feature/feature.xml:5]: This change as was intentional as part of the migration to OH3 and should remain.

**Review by bern77 — CHANGES_REQUESTED** [review]
Hi @ssalonen,

thank you for the changes - looks generally a lot easier to handle 😊

I'll have to test if the suggested changes in `HeliosEasyControlsHandler` work as intended but I think they should. I think I'll wait for the PR to be merged and then test + make adjustments in a separate PR if necessary.

As for the other files, I'm afraid there are many changes which should be part of the `main` branch but which would be reverted to the last `2.5.x` status in your PR?

Thanks and best regards,
Bernhard

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java (hunk 18 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:51]: agreed. I meant that this part of the code should not change at all

**Comment by ssalonen** [comment]
> As for the other files, I'm afraid there are many changes which should be part of the main branch but which would be reverted to the last 2.5.x status in your PR?

As explained above this PR was based on 2.5.x branch first, then auto migrated by openhab GitHub bot to main. 

However, the auto-migration is really not automatic. There are very many changes that are unintentional. I need to quite finely comb through the PR and ensure only the relevant pieces are there, or try finding another way to make this PR on top of main.

**Comment by weymann** [comment]
After changing all versions to 3.0.0 it's fine for me. My E3DC Device is working in the OH3 test environment. 

OH2.5 looks bad because io.transport bundle is missing *net.wimpi.modbus* which I cannot resolve 
```
Error executing command: Error executing command on bundles:
        Error starting bundle 332: Could not resolve module: org.openhab.io.transport.modbus [332]
  Unresolved requirement: Import-Package: net.wimpi.modbus
```

Nevertheless OH3 looks good and I want to approve!

Can you do me one favor and update the README.md for E3DC in order to add the missing pictures:
Please exchange
```
<img style="float: right;" src="doc/GrafanaPV.png">
```
with
```
<img align="right" src="./doc/GrafanaPV.png"/>
```

and
```
<img style="float: right;" src="doc/GrafanaCloudiness.png">
```
with
```
<img align="right" src="./doc/GrafanaCloudiness.png"/>
```

**Comment by ssalonen** [comment]
The relevants changes have been now merged properly on top `main`, original 2.5.x-related conflicts manually resolved.

I ask maintainers to review the relevant changes again please. 

/bundles/org.openhab.binding.modbus.e3dc/ @weymann
/bundles/org.openhab.binding.modbus.helioseasycontrols/ @bern77
/bundles/org.openhab.binding.modbus.stiebeleltron/ @pail23
/bundles/org.openhab.binding.modbus.studer/ @giovannimirulla
/bundles/org.openhab.binding.modbus.sunspec/ @mrbig

**Comment by ssalonen** [comment]
>  Can you do me one favor and update the README.md for E3DC in order to add the missing pictures:

I think it makes sense to separate this to another PR to be honest

**Review by bern77 — APPROVED** [review]
Thanks! LGTM and works with my device.

File: bundles/org.openhab.binding.modbus.e3dc/src/test/java/org/openhab/binding/modbus/e3dc/util/DataConverterTest.java [hunk 26]
```diff
@@ -30,17 +30,58 @@
 public class DataConverterTest {
 
     @Test
-    public void testE3DCSwapValueNegative() {
-        // Reg 69 value 65098 bytes [-2, 74]
-        // Reg 70 value 65535 bytes [-1, -1]
-        byte[] b = new byte[] { -2, -74, -1, -1 };
-        assertEquals(-330, DataConverter.getInt32Swap(ByteBuffer.wrap(b)), "Negative Value");
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus.e3dc/src/test/java/org/openhab/binding/modbus/e3dc/util/DataConverterTest.java:37]: This test not needed anymore?

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+                // Due to nul byte, full register full of nul bytes added
+                .of(new Object[] { "v00020=a", new ModbusRegisterArray(0x7630, 0x3030, 0x3230, 0x3d61, 0x0000) },
+                        new Object[] { "v00020=aa", new ModbusRegisterArray(0x7630, 0x3030, 0x3230, 0x3d61, 0x6100) })
+                .collect(Collectors.toList()));
+    }
+
+    @ParameterizedTest
+    @MethodSource("data")
+    public void testPreparePayload(String payload, ModbusRegisterArray expectedRegisters)
+            throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java:48]: are these throws necessary?

File: bundles/org.openhab.binding.modbus.e3dc/src/main/java/org/openhab/binding/modbus/e3dc/internal/modbus/Parser.java [hunk 28]
```diff
@@ -59,17 +58,10 @@ public Parser(DataType type) {
...
         Optional<ModbusRegisterArray> opt = result.getRegisters();
         if (opt.isPresent()) {
-            ModbusRegisterArray registers = opt.get();
-            int i = 0;
-            for (ModbusRegister reg : registers) {
-                System.arraycopy(reg.getBytes(), 0, newArray, i, 2);
-                i += 2;
-            }
-            setArray(newArray);
+            setArray(opt.get().getBytes());
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus.e3dc/src/main/java/org/openhab/binding/modbus/e3dc/internal/modbus/Parser.java:64]: Are these equivalent?

File: bundles/org.openhab.binding.modbus.e3dc/src/test/java/org/openhab/binding/modbus/e3dc/util/DataConverterTest.java (hunk 26 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.e3dc/src/test/java/org/openhab/binding/modbus/e3dc/util/DataConverterTest.java:37]: Correct. Method under test no longer exist. It was in fact a more convoluted implementation of a conversion that exists in transport module.

While this test has been added, you should be able to find the same "byte sequence", representing negative value, tested in a new test introduced in data block test.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java (hunk 27 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java:48]: These are from the reflection, so I believe they are? Alternatively I could try catch and fail/throw unchecked exception

File: bundles/org.openhab.binding.modbus.e3dc/src/main/java/org/openhab/binding/modbus/e3dc/internal/modbus/Parser.java (hunk 28 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.e3dc/src/main/java/org/openhab/binding/modbus/e3dc/internal/modbus/Parser.java:64]: Yes, that should be equivalent. Tests should further give confirmation on this.

Previously it was hard to get raw bytes "abstracted" by ModbusRegisterArray and this type of copying of bytes was needed. 

A bit of modbus terminology: register means 16 bit (big endian) word returned by the modbus device. You could consider modbus being a remotely accessible memory, memory locations represented by register addresses.

Often in these bindings the register data is converted to 16/32 bit integers and floats, that we give meaning to that raw data.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java (hunk 27 above)
Comment (fwolter) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java:48]: The checked exceptions `IllegalAccessException` and `InvocationTargetException` are never thrown. So they could be removed. Specifying an unchecked exception like `IllegalArgumentException` has no effect in tests, AFAIK. So, that could be removed, too.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java (hunk 27 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/test/java/org/openhab/binding/modbus/helioseasycontrols/internal/PreparePayloadTest.java]: Resolved in e43bdc610

**Comment by ssalonen** [comment]
Rebased on top of `main` to resolve jenkins issues with some of the bluetooth null annotations.

**Comment by ssalonen** [comment]
[Jenkins](https://ci.openhab.org/job/PR-openHAB-Addons/988/): ```
[ERROR] Failed to execute goal on project org.openhab.binding.tado: Could not resolve dependencies for project org.openhab.addons.bundles:org.openhab.binding.tado:jar:3.0.0-SNAPSHOT: Failure to find org.openhab.binding.tado:api-client:jar:1.4.1 in https://openhab.jfrog.io/openhab/libs-release was cached in the local repository, resolution will not be reattempted until the update interval of openhab-release has elapsed or updates are forced -> [Help 1]
```

**Comment by cpmeister** [comment]
modbus integration tests are failing

**Comment by ssalonen** [comment]
In Jenkins the modbus integration tests are skipped though?

[INFO] openHAB Add-ons :: Integration Tests :: Modbus Binding Tests SKIPPED

I think the reason is the dependency errors with tado binding (commented above). Is this known?

**Comment by fwolter** [comment]
The build fails because the header in org.openhab.binding.modbus.helioseasycontrols.internal.PreparePayloadTest is missing and org.openhab.binding.modbus.e3dc.dto.DataBlockTest.testValidPowerBlockNegativePVSupply failed.

**Comment by ssalonen** [comment]
Thank you very much @fwolter, I don't know why I missed that...was focusing too much in the last error of the jenkins output.

Commits added to resolve the issues highlighted, let's see.

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**